Update ML-IAP for ACE support

jmgoff opened the pull request:
**Summary**

<!--Briefly describe the new feature(s), enhancement(s), or bugfix(es) included in this pull request.-->
The ML-IAP package is updated to support ACE descriptors and models. Previously ML-IAP could not support ACE descriptor computes or ACE models (including linear, neural network, and more general non-linear models).  

Key enhancements and features are:

- LAMMPS-native support for arbitrary ACE pytorch models (including neural networks)
-- Users can run arbitrary ACE neural network poentials in LAMMPS, calculate energies, forces, and stresses
- Ability for users to separate linear model coefficients and clebsch-gordan coefficients (offers a more general format for ACE models developed from different software)
-- This may help serve as a template to simplify ACE inputto a single list of coefficients in the future
- Ability to use ACE descriptors in general mliappy/python routines (e.g. in structure generation algorithms, model training, and on-the-fly analysis)
- Various examples for ML-IAP + ACE are provided


**Related Issue(s)**

N/A
<!--If this addresses an open GitHub issue for this project, please mention the issue number here, and describe the relation. Use the phrases `fixes #221` or `closes #135`, when you want an issue to be automatically closed when the pull request is merged-->

**Author(s)**

<!--Please state name and affiliation of the author or authors that should be credited with the changes in this pull request. If this pull request adds new files to the distribution, please also provide a suitable "long-lived" e-mail address (ideally something that can outlive your institution's e-mail, in case you change jobs) for the *corresponding* author, i.e. the person the LAMMPS developers can contact directly with questions and requests related to maintenance and support of this contributed code.-->
James Goff (Sandia National Laboratories) ( [goffjimmy1@gmail.com](mailto:goffjimmy1@gmail.com) )

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Will not break backward compatibility 
<!--Please state whether any changes in the pull request will break backward compatibility for inputs, and - if yes - explain what has been changed and why-->

**Implementation Notes**

<!--Provide any relevant details about how the changes are implemented, how correctness was verified, how other features - if any - in LAMMPS are affected-->
ACE support in mliap uses functions from the ML-PACE package, and updates were made to import new ACE descriptor evaluators implemented by the owners of the ML-PACE package. This largely required the development of a new mliap descriptor class in`src/ML-IAP/mliap_descriptor_ace.*` - and minor adjustments throughout the mliap package to accommodate the new class.

To validate the results for the mliap implementation of ACE descriptors and models:
The mliap ace compute reproduces descriptor values, descriptor gradients, and virial terms from`compute pace`. 
The energies, forces, and stresses were validated for mliap ACE models (including pytorch neural networks)  the `numdiff` fix.


**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

<!--Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)-->

jmgoff committed:
ACE mliap initial commit

jmgoff committed:
updated examples

jmgoff committed:
Merge branch 'lammps:develop' into pace-mliap

jmgoff committed:
ACE mliap examples and virial support

jmgoff committed:
cleanup and whitespace

jmgoff committed:
ACE mliap examples

jmgoff committed:
added mliap_enable_ace to opt in/out of ace functionality in mliap

jmgoff commented:
Similar to the ML-IAP  python models requiring that PKG_PYTHON is installed, the ML-IAP ACE descriptors require that PKG_ML-PACE is also installed. This crashed some unit tests before because ML-PACE is not included by default.

Resolved by adding a new compilation flag: `-DMLIAP_ENABLE_ACE` that can only be turned on if `-DPKG_ML-PACE` is on.

To use ML-IAP & ACE functionality, compile lammps with the following dflags (in addition to the dflags required for Python and ML-IAP:
`-DPKG_ML-PACE=yes`
`-DMLIAP_ENABLE_ACE=yes`

akohlmey reviewed (changes requested):
This will need updates to the documentation.
Also, I only see support for the CMake build system. What about the traditional build with GNU make?

akohlmey commented:
Since both ML-IAP and ML-PACE have KOKKOS versions and also mliappy has KOKKOS support, does this PR need some KOKKOS updates?

athomps commented:
> Since both ML-IAP and ML-PACE have KOKKOS versions and also mliappy has KOKKOS support, does this PR need some KOKKOS updates?

The same functionality already exists for SNAP descriptors, but these have not yet been Kokkosized in MLIAP.  I think this will happen soon, but until then, it is better not to attempt it for ACE descriptors, which are quite a bit more complicated than the SNAP descriptors

akohlmey commented on cmake/Modules/Packages/ML-IAP.cmake:
```diff
@@ -10,6 +10,15 @@ endif()
+  set(MLIAP_ENABLE_PYTHON_DEFAULT ON)
+endif()
+
+option(MLIAP_ENABLE_PYTHON "Build ML-IAP package with Python support" ${MLIAP_ENABLE_PYTHON_DEFAULT})
```
This is replicated from line 11

akohlmey commented on src/ML-IAP/mliap_descriptor_ace.cpp:
```diff
@@ -0,0 +1,432 @@
+
+using namespace LAMMPS_NS;
+
+#define MAXLINE 1024
```
These two defines are not used anywhere.

akohlmey commented:
> but until then, it is better not to attempt it for ACE descriptors

Then there should be some tests implemented that prevent using this with KOKKOS or someone needs to explain how this cannot produce problems with the current code if somebody tries this.

akohlmey commented on src/ML-IAP/mliap_descriptor_ace.cpp:
```diff
@@ -0,0 +1,432 @@
+#include <cmath>
+#include <cstring>
+#include <typeinfo>
+//#include <map>
```
Please remove commented out code like this.

akohlmey commented on src/ML-IAP/mliap_descriptor_ace.cpp:
```diff
@@ -0,0 +1,432 @@
+
+#include <cmath>
+#include <cstring>
+#include <typeinfo>
```
Is this include file needed?

stanmoore1 commented:
I am OK with Aidan's suggestion of waiting on adding Kokkos support for ACE/SNAP. So I think this PR can be merged after addressing Axel's feedback.

jmgoff committed:
mliap ace docs and example logs

akohlmey committed:
Merge branch 'develop' into pace-mliap

# Conflicts:
#	src/ML-IAP/compute_mliap.cpp

akohlmey commented:
@jmgoff FYI, I've merged this branch with upstream and resolved a merge conflict due to recently added functionality.

jmgoff committed:
Merge branch 'lammps:develop' into pace-mliap

athomps reviewed (approved):
I approve

stanmoore1 reviewed (changes requested):
Need to error out if a user tries to use the Kokkos version of ML-IAP with ACE or SNAP descriptors, something like this:

```
if (lmp->kokkos) error->all(FLERR,"Cannot (yet) use KOKKOS package with SNAP descriptors");
```

jmgoff committed:
kokkos error and standard make

jmgoff committed:
Merge branch 'lammps:develop' into pace-mliap

jmgoff committed:
fixed whitespace in docs

jmgoff committed:
Merge branch 'pace-mliap' of github.com:jmgoff/lammps_compute_PACE into pace-mliap

jmgoff committed:
cleanup

jmgoff commented:
Thank you all for the comments and reviews. The most recent pushes should address the remaining comments and issues. To summarize:
For Axel's previous comments:
-Documentation has been updated and added for ace in MLIAP
-Cleaned up leftover comments, defines, and whitespace in docs
-Support for make added. Updated ML-IAP make file setup from `ML-IAP` to allow for optional ace descriptor use (requiring that ML-PACE is also built) 
For Aidan's comments:
-Standard g++ (1 core) and (4 core) outputs have been added for the new MLIAP/ace examples
To address comments from Axel, Stan, and Aidan on the use of kokkos:
-Following Stan's comment this morning and previous discussion, checks have been added to `pair_mliap.cpp` and `compute_mliap.cpp` to throw runtime errors when users attempt to use `mliap/kk` with SNAP or ACE in mliap computes or pair styles.

akohlmey committed:
silence compiler warnings and apply clang-format

akohlmey committed:
correct CMake script logic

akohlmey committed:
correct and complete conventional build package logic

akohlmey committed:
apply clang-format

akohlmey commented:
@jmgoff both CMake and GNU make support were incorrect or incomplete. I have tried to fix it.

There were still some style issues that I resolved and I also applied clang-format to make the programming style consistent. I also reordered the initializer list of the MLIAPData class to avoid compiler warnings.

Since CPU snap descriptors are no longer allowed for KOKKOS, they corresponding unit tests need to be skipped.

akohlmey committed:
must skip KOKKOS tests for any ML-IAP inputs using snap descriptors

akohlmey committed:
fix typo

jmgoff commented:
> @jmgoff both CMake and GNU make support were incorrect or incomplete. I have tried to fix it.
> 
> There were still some style issues that I resolved and I also applied clang-format to make the programming style consistent. I also reordered the initializer list of the MLIAPData class to avoid compiler warnings.
> 
> Since CPU snap descriptors are no longer allowed for KOKKOS, they corresponding unit tests need to be skipped.

Hi Axel, thanks for applying the extra corrections for the CMake and gmake scripts along with the styling issues. I'll keep an eye out for similar things in the future.

As long as you and @stanmoore1 are okay with disabling mliap/kk + CPU snap for now, this sounds good to me.

stanmoore1 commented:
Does it need to be disabled for Kokkos? i.e. does it give the wrong answer, or is it just slow but still works correctly? I thought it was giving the wrong answer or couldn't run, but if it is just slow then maybe we just keep it?

jmgoff commented:
That is a good point. SNAP is only crashes for the mliappy model, not for the linear, quadratic, or NN models. I'll add in errors for the mliappy model specifically to make sure that we don't have to turn off unit tests that work.

jmgoff committed:
mliap kk snap unittest back on  only disable mliap kk for ACE

jmgoff commented:
Adjusted it to only have ACE + mliap kokkos disabled and turned unit tests back on for SNAP + mliap kokkos. The SNAP mliappy model example works as well with the SNAP CPU descriptors.

akohlmey committed:
Merge branch 'develop' into pace-mliap

akohlmey committed:
flag false positives

akohlmey commented:
@stanmoore1 this is now waiting on you

stanmoore1 commented on doc/src/pair_mliap.rst:
```diff
@@ -169,6 +198,24 @@ contains a few more arguments (e.g., *nmax* and *alpha*). The preparation
+embedding ACE potentials used in the :doc:`pair_style pace <pair_pace>`.
+As noted above, the key difference is that the Clebsch-Gordan coefficients
+in the descriptor file with *mliap descriptor ace* are *NOT* multiplied
+multiplied by linear or square root embedding terms. In other words,
```
Typo: "multiplied multiplied" (occurs twice)

stanmoore1 commented on doc/src/pair_mliap.rst:
```diff
@@ -72,6 +73,24 @@ the smooth SO(3) power spectrum with the explicit inclusion of a radial
+ML-IAP are *linear* and *mliappy*. The *ace* descriptors and models require
+building LAMMPS with the ML-PACE package (see below). The *mliappy* model
+style may be used with *ace* descriptors, but it requires that LAMMPS is
+also built with the PYTHON package. As with other, the *mliappy* model style
```
"As with other," ... not sure what this means or what "other" refers to?

stanmoore1 reviewed (approved):
A few minor doc issues, otherwise looks good.

akohlmey committed:
small doc fixes

akohlmey committed:
breathe is currently not compatible with sphinx 7.3
